models/repo/setting.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repo
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"code.gitea.io/gitea/models/db"
+
+	"xorm.io/builder"
+)
+
+// Setting is a key value store of repo settings
+type Setting struct {
+	ID           int64  `xorm:"pk autoincr"`
+	RepoID       int64  `xorm:"index unique(key_repoid)"`              // to load all of someone's settings
+	SettingKey   string `xorm:"varchar(255) index unique(key_repoid)"` // ensure key is always lowercase
+	SettingValue string `xorm:"text"`
+}
+
+// TableName sets the table name for the settings struct
+func (s *Setting) TableName() string {
+	return "repo_setting"
+}
+
+func init() {
+	db.RegisterModel(new(Setting))
+}
+
+// GetRepoSettings returns specific settings from repo
+func GetRepoSettings(rid int64, keys []string) (map[string]*Setting, error) {
+	settings := make([]*Setting, 0, len(keys))
+	if err := db.GetEngine(db.DefaultContext).
+		Where("repo_id=?", rid).
+		And(builder.In("setting_key", keys)).
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	settingsMap := make(map[string]*Setting)
+	for _, s := range settings {
+		settingsMap[s.SettingKey] = s
+	}
+	return settingsMap, nil
+}
+
+// GetRepoAllSettings returns all settings from repo
+func GetRepoAllSettings(rid int64) (map[string]*Setting, error) {
+	settings := make([]*Setting, 0, 5)
+	if err := db.GetEngine(db.DefaultContext).
+		Where("repo_id=?", rid).
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	settingsMap := make(map[string]*Setting)
+	for _, s := range settings {
+		settingsMap[s.SettingKey] = s
+	}
+	return settingsMap, nil
+}
+
+func validateRepoSettingKey(key string) error {
+	if len(key) == 0 {
+		return fmt.Errorf("setting key must be set")
+	}
+	if strings.ToLower(key) != key {
+		return fmt.Errorf("setting key should be lowercase")
+	}
+	return nil
+}
+
+// GetRepoSetting gets a specific setting for a repo
+func GetRepoSetting(repoID int64, key string, def ...string) (string, error) {
+	if err := validateRepoSettingKey(key); err != nil {
+		return "", err
+	}
+	setting := &Setting{RepoID: repoID, SettingKey: key}
+	has, err := db.GetEngine(db.DefaultContext).Get(setting)
+	if err != nil {
+		return "", err
+	}
+	if !has {
+		if len(def) == 1 {
```

Comment:
```suggestion
		if len(def) >= 1 {
```
Or should the default value not be supplied if you add too many arguments?

Reply:
The question is why you should add too many arguments? What's the meaning of it ....

`>=` doesn't make the logic more correct. If you really want to ensure the correctness, you should do:

```
if len(def) == 1 {
    return def[0]
} else if len(def) > 1 {
    panic()
}
```